src/core/middlewares/regquota/handler.go

```diff
@@ -37,38 +55,173 @@ func New(next http.Handler) http.Handler {
-	match, _, _ := util.MatchManifestURL(req)
+// ServeHTTP manifest ...
+func (rqh *regQuotaHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	match, repository, tag := util.MatchPushManifest(req)
 	if match {
-		var mfSize int64
-		var mfDigest string
+		mfInfo := &util.MfInfo{
+			Repository: repository,
+			Tag:        tag,
+		}
+		rqh.mfInfo = mfInfo
+
 		mediaType := req.Header.Get("Content-Type")
-		if req.Method == http.MethodPut {
-			if mediaType == schema1.MediaTypeManifest ||
-				mediaType == schema1.MediaTypeSignedManifest ||
-				mediaType == schema2.MediaTypeManifest {
-				data, err := ioutil.ReadAll(req.Body)
-				if err != nil {
-					log.Warningf("Error occurred when to copy manifest body %v", err)
-					http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to decode manifest body %v", err)), http.StatusInternalServerError)
-					return
-				}
-				req.Body = ioutil.NopCloser(bytes.NewBuffer(data))
+		if mediaType == schema1.MediaTypeManifest ||
+			mediaType == schema1.MediaTypeSignedManifest ||
+			mediaType == schema2.MediaTypeManifest {
+
+			tagLock, err := rqh.tryLockTag()
+			if err != nil {
+				log.Warningf("Error occurred when to lock tag %s:%s with digest %v", repository, tag, err)
+				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to lock tag %s:%s with digest %v", repository, tag, err)), http.StatusInternalServerError)
+				return
+			}
+			rqh.mfInfo.TagLock = tagLock
+
+			data, err := ioutil.ReadAll(req.Body)
+			if err != nil {
+				rqh.tryFreeTag()
+				log.Warningf("Error occurred when to copy manifest body %v", err)
+				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to decode manifest body %v", err)), http.StatusInternalServerError)
+				return
+			}
+			req.Body = ioutil.NopCloser(bytes.NewBuffer(data))
 
-				_, desc, err := distribution.UnmarshalManifest(mediaType, data)
-				if err != nil {
-					log.Warningf("Error occurred when to Unmarshal Manifest %v", err)
-					http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to Unmarshal Manifest %v", err)), http.StatusInternalServerError)
-					return
+			manifest, desc, err := distribution.UnmarshalManifest(mediaType, data)
+			if err != nil {
+				rqh.tryFreeTag()
+				log.Warningf("Error occurred when to Unmarshal Manifest %v", err)
+				http.Error(rw, util.MarshalError("InternalServerError", fmt.Sprintf("Error occurred when to Unmarshal Manifest %v", err)), http.StatusInternalServerError)
+				return
+			}
+			rqh.mfInfo.Refrerence = manifest.References()
+			rqh.mfInfo.Digest = desc.Digest.String()
+			rqh.mfInfo.Size = desc.Size
+
+			projectID, err := rqh.getProjectID(strings.Split(repository, "/")[0])
+			if err != nil {
+				log.Warningf("Error occurred when to get project ID %v", err)
+				return
+			}
+			rqh.mfInfo.ProjectID = projectID
+
+			exist, af, err := rqh.imageExist()
```

Comment:
Should split the logic of handling storage quota and number quota here

Reply:
yes, move the logic on size to the blob handler.